app/src/server/api/routers/externalApi.router.ts

```diff
@@ -0,0 +1,205 @@
+        reqPayload: z.unknown().describe("JSON-encoded request payload"),
+        tags: z
+          .record(z.string())
+          .optional()
+          .describe(
+            'Extra tags to attach to the call for filtering. Eg { "userId": "123", "promptId": "populate-title" }',
+          ),
+      }),
+    )
+    .output(
+      z.object({
+        respPayload: z.unknown().optional().describe("JSON-encoded response payload"),
+      }),
+    )
+    .mutation(async ({ input, ctx }) => {
+      const apiKey = ctx.apiKey;
+      if (!apiKey) {
+        throw new TRPCError({ code: "UNAUTHORIZED" });
+      }
+      const key = await prisma.apiKey.findUnique({
+        where: { apiKey },
+      });
+      if (!key) {
+        throw new TRPCError({ code: "UNAUTHORIZED" });
+      }
+      const reqPayload = await reqValidator.spa(input.reqPayload);
+      const cacheKey = hashRequest(key.organizationId, reqPayload as JsonValue);
+
+      const existingResponse = await prisma.loggedCallModelResponse.findFirst({
+        where: {
+          cacheKey,
+        },
+        include: {
+          originalLoggedCall: true,
+        },
+        orderBy: {
+          startTime: "desc",
+        },
+      });
+
+      if (!existingResponse) return { respPayload: null };
+
+      await prisma.loggedCall.create({
+        data: {
+          organizationId: key.organizationId,
+          startTime: new Date(input.startTime),
+          cacheHit: true,
+          modelResponseId: existingResponse.id,
+        },
+      });
+
+      return {
+        respPayload: existingResponse.respPayload,
+      };
+    }),
+
+  report: publicProcedure
+    .meta({
+      openapi: {
+        method: "POST",
+        path: "/v1/report",
+        description: "Report an API call",
+      },
+    })
+    .input(
+      z.object({
+        startTime: z.number().describe("Unix timestamp in milliseconds"),
+        endTime: z.number().describe("Unix timestamp in milliseconds"),
+        reqPayload: z.unknown().describe("JSON-encoded request payload"),
+        respPayload: z.unknown().optional().describe("JSON-encoded response payload"),
+        respStatus: z.number().optional().describe("HTTP status code of response"),
+        error: z.string().optional().describe("User-friendly error message"),
+        tags: z
+          .record(z.string())
+          .optional()
+          .describe(
+            'Extra tags to attach to the call for filtering. Eg { "userId": "123", "promptId": "populate-title" }',
+          ),
+      }),
+    )
+    .output(z.void())
+    .mutation(async ({ input, ctx }) => {
+      const apiKey = ctx.apiKey;
+      if (!apiKey) {
+        throw new TRPCError({ code: "UNAUTHORIZED" });
+      }
+      const key = await prisma.apiKey.findUnique({
+        where: { apiKey },
+      });
+      if (!key) {
+        throw new TRPCError({ code: "UNAUTHORIZED" });
+      }
+      const reqPayload = await reqValidator.spa(input.reqPayload);
```

Comment:
I'm still concerned about this. If someone messes up their prompt construction and accidentally calls OpenAI with `messages` set to `null`, my understanding is this will just throw a server error they'll never see and the log will never show up. Wouldn't it be better to log the request anyway, along with the error response they got from OpenAI, so they know what went wrong?

Reply:
Oh I see what you mean. Yes, let's change this to no longer require proper validation.